nengo/builder/tests/test_optimizer.py

```diff
@@ -0,0 +1,119 @@
+import numpy as np
+import pytest
+
+from nengo.builder.optimizer import (
+    compatible, check_signals_mergeable, check_views_mergeable, merge_signals,
+    merge_views)
+from nengo.builder.signal import Signal
+
+
+def test_compatible():
+    # 0-d signals
+    assert compatible([Signal(0), Signal(0)])
+    assert not compatible([Signal(0), Signal(1)])
+
+    # compatible along first axis
+    assert compatible(
+        [Signal(np.empty((1, 2))), Signal(np.empty((2, 2)))])
+
+    # compatible along second axis
+    assert compatible(
+        [Signal(np.empty((2, 1))), Signal(np.empty((2, 2)))], axis=1)
+    assert not compatible(
+        [Signal(np.empty((2, 1))), Signal(np.empty((2, 2)))], axis=0)
+
+    # shape mismatch
+    assert not compatible(
+        [Signal(np.empty((2,))), Signal(np.empty((2, 2)))])
+
+    # mixed dtype
+    assert not compatible(
+        [Signal(np.empty(2, dtype=int)), Signal(np.empty(2, dtype=float))])
```

Comment:
It definitely has the potential to change the exact model results. Obviously, we can't convert the `float` to `int`. Converting `int` to `float` opens the door to rounding errors and all kind of weird behaviours. If we are using `int` signals somewhere, we probably would not want to convert them to `float`.